File: 7-kyu/anagram-detection/c++/solution.cpp
#include <algorithm>
#include <cctype>
#include <map>
#include <string>

bool isAnagram(std::string test, std::string original)
{
    std::transform(test.begin(), test.end(), test.begin(), ::tolower);
    std::transform(original.begin(), original.end(), original.begin(), ::tolower);
    std::map<char, int> test_counts, original_counts;
    for (const auto &c : test) {
        test_counts[c]++;
    }
    for (const auto &c : original) {
        original_counts[c]++;
    }
    return test_counts == original_counts;
}

File: 7-kyu/anagram-detection/javascript/solution.js
const isAnagram = (test, original) => {
  const testLower = test.toLowerCase();
  const origLower = original.toLowerCase();
  const testCounts = {};
  const origCounts = {};
  for (const letter of testLower) {
    testCounts[letter] = testCounts[letter] + 1 || 1;
  }
  for (const letter of origLower) {
    origCounts[letter] = origCounts[letter] + 1 || 1;
  }
  if (Object.entries(origCounts).length !== Object.entries(testCounts).length) {
    return false;
  }
  for (const [ letter, count ] of Object.entries(origCounts)) {
    if (count !== testCounts[letter]) {
      return false;
    }
  }
  return true;
};
